services/src/main/java/org/keycloak/authentication/requiredactions/UpdateProfile.java

```diff
@@ -73,36 +72,34 @@ public void processAction(RequiredActionContext context) {
         String oldFirstName = user.getFirstName();
         String oldLastName = user.getLastName();
         String oldEmail = user.getEmail();
```

Comment:
As I commented in UserProfile already, these old values should be propagated into update listener by profile.update() not to be handled this way

Reply:
I did not want to add a new type but keep the API simpler.

Also, I'm not sure if that code in particular should be handled by client code, but within the user profile itself.